### DISCUSSION TITLE
Where to insert a Resource needed for a Plugin when expansive computation is required?

### DISCUSSION THREAD

**Discussion by IceSentry** [bevyengine/bevy#2519] [Q&A] [answered]
If my plugin adds a Resource that needs to do a few somewhat intensive things before it's ready to be added, should I insert it in a startup system or directly in the build function of a Plugin?

**Comment by will-hart** [comment, marked as answer]
I'm not sure what the "official" approach is, but I would add the resource in an "empty" state in the build function. I would then have the expensive setup process work in a separate, earlier game state to the system(s) that consume it. Depending on the expensiveness of your expensive things you may also want to consider running the expensive things using [bevy's async features](https://github.com/bevyengine/bevy/blob/main/examples/async_tasks/async_compute.rs).

For example a sync version might look something like this:

```rust
// in plugin builder
app.insert_resource(ExpensiveComputationForThis::default())
    .add_system_set(SystemSet::on_enter(MyGameState::DoExpensiveThings).with_system(do_expensive_things.system())
    .add_system_set(SystemSet::on_update(MyGameState:AfterDoingExpensiveThings).with_system(use_expensive_things.system());

// in the `do_expensive_things` system
fn do_expensive_things(mut expensive: ResMut<ExpensiveComputationForThis>, mut state: ResMut<State<MyGameState>>) {
    // do something expensive
    expensive.gym_membership = exercise_expensively();

    // move to the next game state
    state.set(MyGameState::AfterDoingExpensiveThings).unwrap();
}
```

**Reply by IceSentry** [reply]
I'm essentially doing something like that right now, but I only insert the Resource in the system once it's ready. Any reason why you insert it then mutate in the system it instead of just generating it in the system?

**Reply by will-hart** [reply]
No real reason, just for consistency - i.e. I add all my resources at the start of the plugin it belongs to, which I think is easier and for me less likely to break something than having resources inserted in various parts of the code base in various systems / stages etc.

**Comment by ManevilleF** [comment]
You can use the state system and make your plugin struct generic:

```rust
struct MyPlugin<T> {
     pub start_state: T
}

impl<T: 'static + Debug + Clone + Eq + PartialEq + Hash + Send + Sync> Plugin for MyPlugin<T> {
    fn build(&self, app: &mut AppBuilder) {
        app.add_system_set(
            SystemSet::on_enter(self.start_state.clone())
                .with_system(Self::my_expensive_system.system()),
        )
    }
}
```

**Reply by IceSentry** [reply]
I'm not sure I understand the point of making this generic. I'm also not sure where should I make the `.insert_resouce()` call.

**Reply by ManevilleF** [reply]
You can avoid using startup systems with states
